File: packages/rust-printer/src-ts/print.ts
import { PrintItemIterable } from "@dprint/types";
import * as rustPrinter from "./pkg/dprint_rust_printer";
import { preparePrintItems } from "./preparePrintItems";
import { PrintOptions } from "./PrintOptions";
import { printWriteItems } from "./printWriteItems";

/**
 * Print out the provided print items using the rust printer.
 * @param items - Items to print.
 * @param options - Options for printing.
 */
export function print(items: PrintItemIterable, options: PrintOptions) {
    const writeItems = rustPrinter.get_write_items(preparePrintItems(items), options.maxWidth, options.indentWidth, options.isTesting);
    return printWriteItems(writeItems, options);
}

File: packages/rust-core/src/print.rs
use super::*;

/// Options for printing the print items.
pub struct PrintOptions {
    /// The width the printer will attempt to keep the line under.
    pub max_width: u32,
    /// The number of columns to count when indenting or using a tab.
    pub indent_width: u8,
    /// Whether to use tabs for indenting.
    pub use_tabs: bool,
    /// The newline character to use when doing a new line.
    pub new_line_text: &'static str,
}

/// Prints out the print items using the provided
pub fn print<TString, TInfo, TCondition>(
    print_items: PrintItems<TString, TInfo, TCondition>,
    options: PrintOptions
) -> String where TString : StringTrait, TInfo : InfoTrait, TCondition : ConditionTrait<TString, TInfo, TCondition> {
    let write_items = get_write_items(&print_items, GetWriteItemsOptions {
        indent_width: options.indent_width,
        max_width: options.max_width,
    });

    print_write_items(write_items, PrintWriteItemsOptions {
        use_tabs: options.use_tabs,
        new_line_text: options.new_line_text,
        indent_width: options.indent_width,
    })
}
